Cleaning

Benj1bear opened the pull request:
- Fixed slicing
 - Removed unnecessary imports and code
 - Added the patch_iterators initialization to custom_generator.py
 - Fixed scope error in sign function in source_processing.py 
 - Fixed uninitialized iteration behaviour

All tests still work without pytest it's just that pytest adds stuff to the ```globals()``` that messes with tests that are reliant on ```globals()``` i.e. some tests are failing from ```globals()``` or ```inspect.currentframe()``` most likely.

With pytest running I get:

![image](https://github.com/user-attachments/assets/43d007ea-820f-48b8-8600-0435908bd299)


I'm thinking about how to scope the patches to avoid assigning them into globals but the problem you encounter if you do so is that we're essentially wanting unbound locals to be dynamically defined which I don't think python lets us do because it will compile any unbound locals that are not assigned as being globals/builtins/ or nonlocals/closures if these are defined.

I.e.:
```python
import ctypes
from inspect import currentframe
from gcopy.track import patch_iterators

def test():
    range = 3  # we don't want this line here but it's needed at compile time (I think) to indicate its scoping
    local_frame = currentframe()
    patch_iterators(local_frame.f_locals)
    # only needed in python versions before PEP 667 implementation (I think)
    # e.g. when locals was only a static view rather than the live view
    ctypes.pythonapi.PyFrame_LocalsToFast(ctypes.py_object(local_frame), ctypes.c_int(0))
    for i in range(3):
        print(locals()[".internals"][".4"])
test()
```

Benj1bear committed:
Update test_custom_generator.py

Fixed slicing
Removed unnecessary imports and code
Added the patch_iterators initialization to custom_generator.py
Fixed scope error in sign function in source_processing.py
Fixed uninitialized iteration behaviour
Fixed yield from lineno
Made patch_iterators scoped
updated async tests based on the changes

Benj1bear commented:
I just noticed a small mistake. Need to update lineno for ```yield from```

Benj1bear commented:
I found one solution to the scoping problem:

```python
from types import FunctionType

patched_iterators = {"range": 4}

def patch_iterators_locally(FUNC: FunctionType) -> FunctionType:
    return FunctionType(FUNC.__code__, globals() | patched_iterators, FUNC.__name__, FUNC.__defaults__, FUNC.__closure__)


@patch_iterators_locally
def test() -> None:
    print(range)

test()
print(range, globals().get("range"))
```

It seems to work fine for classes though:

```python
from gcopy.track import patch_iterators

class j:
    patch_iterators()
    print(type(range))
print(type(range))
```
Users will only need this if they want to initialize from a running generator with implicit iterators preserved beforehand and they don't want the patches in the global scope.

I'll add it in just in case users want this. It will mean they have three use cases for the ```patch_iterators``` function from track.py for when they decide to use already running generators instead of for whatever reason not wanting to start with our Generator or AsyncGenerator types:

```python
## globally ##
patch_iterators()

## only for the functions scope ##
@patch_iterators
def test():
    ...

## only for the classes scope ##
class test:
    patch_iterators()
```

PabloRuizCuevas commented:
This looks great, you tell me if is ready to merge or for me to test the changes.

Benj1bear commented:
cool, I'm just going to review some more of it and then add examples, after that it should be ready and I'll let you know.

Benj1bear committed:
Update custom_generator.py

Benj1bear committed:
Update custom_generator.py

Fixed value yields that are without brackets

Benj1bear committed:
Update source_processing.py

small cleaning
+
replaced code around specific iterator checking (is_running, get_iter_index, and SetIteratorView) to frame_state checking (inspect.getgeneratorstate) since this is much better
Fixed extract_function
Fixed extract_source_from_comparison
Fixed bracketing in unpack but need to test further to verify

Updated TODO

Benj1bear committed:
Update source_processing.py

small cleaning
+
Fixed bracketing issues around unpacking yields but need to test to further verify
Changed some of the exception formats
Fixed async tests for pytest
Changed getgeneratorstate to a custom get_state for generality
Updated TODO

Benj1bear committed:
Update source_processing.py

small cleaning
+
Added tests for brackets for clean_source_lines
Added tests for except + ExceptionGroups + 3.14 change for non-bracketed expression in the except
Fixed some of the tests failing due to pytest specifics
Fixed nested exceptions

Updated TODO
